etw/cs/traceloggingdynamic/TraceLoggingDynamic.cs

```diff
@@ -1604,83 +1763,86 @@ internal EventProvider.EventRawInfo GetRawData()
             };
         }
 
-        private void AddScalarMetadata(string name, InType inType, EventOutType outType, int tag)
+        /// <returns>The position of the outType byte within the metadata array.</returns>
+        private unsafe int AddMetadata(string name, InType inType, EventOutType outType, int tag)
         {
+            int metadataPos;
             Debug.Assert(name.IndexOf('\0') < 0, "Field name must not have embedded NUL characters.");
             Debug.Assert(((int)outType & 127) == (int)outType, "Invalid outType.");
             Debug.Assert((tag & 0xfffffff) == tag, "Tag must be 28-bit value.");
 
+            int pos;
+
+            var nameLength = name.Length;
+            var nameMaxByteCount = this.utf8NameEncoding.GetMaxByteCount(nameLength);
             if (tag != 0)
             {
-                int pos = this.AddMetadataName(name, 7);
-                this.metadata.data[pos++] = 0; // nul
-                this.metadata.data[pos++] = (byte)((byte)inType | ChainBit);
-                this.metadata.data[pos++] = (byte)((byte)outType | ChainBit);
-                this.metadata.data[pos++] = unchecked((byte)(0x80 | tag >> 21));
-                this.metadata.data[pos++] = unchecked((byte)(0x80 | tag >> 14));
-                this.metadata.data[pos++] = unchecked((byte)(0x80 | tag >> 7));
-                this.metadata.data[pos++] = unchecked((byte)(0x7F & tag));
-                this.metadata.SetSize(pos);
+                pos = this.metadata.ReserveSpaceFor(nameMaxByteCount + 7);
+                fixed (char* pName = name)
+                {
+                    fixed (byte* pMeta = this.metadata.data)
+                    {
+                        pos += this.utf8NameEncoding.GetBytes(pName, nameLength, pMeta + pos, nameMaxByteCount);

```

Comment:
Note: Inlined 3 copies of AddMetadataName into AddMetadata, and used unsafe code to avoid redundant array bounds checks. Small but noticeable performance improvement.